### PR TITLE
A small certificate is good

### LINKED ISSUE #3784: Improve Large Server Certificate Scenario
### Problem

Amplification protection (send no more than 3x of receive) on the server side ends up adding an extra round trip in many scenarios than use a large server certificate.

### Proposed Solution

Send more in the initial client flight.

We already require client Initial packets to be padded to at least 1200 bytes to force them to "pay to play" so to speak, but that's not enough if the server's certificate is larger than ~3K. [This blog post](https://www.fastly.com/blog/quic-handshake-tls-compression-certificates-extension-study) does a good job of outlining the problem and proposes certificate compression as a possible solution.

A simpler solution (IMO) is to recommend clients send more in their initial flight; perhaps two fully padded copies of their Client Hello. This effectively doubles the size of certificate that would fit in the server's first flight.

Doubling the number of initial packets may increase congestion at the start of the handshake, but QUIC has already made the decision to send 2 probe packets for recovery, compared to TCP's 1.

### PR DESCRIPTION
**Pull request by martinthomson**
There are a few ways that this can be managed, but it's worth pointing
out why, and some things that might help.

Closes #3784.

### CODE REVIEW

File: draft-ietf-quic-tls.md [hunk 1]
```diff
@@ -651,6 +651,17 @@ verification that the identity of the server is included in a certificate and
...
 {{?RFC2818}}).
 
+Note:
+
+: Where servers provide certificates for authentication, the size of
+  the certificate chain can consume a large number of bytes.  Controlling the
+  size of certificate chains is critical to performance in QUIC as servers are
+  limited to sending 3 bytes for every byte received prior to validating the
+  client address; see Section 8.1 of {{QUIC-TRANSPORT}}.  The size of a
+  certificate chain can managed by limiting the number of names or extensions;
```
Comment (marten-seemann) [inline draft-ietf-quic-tls.md:661]: can be?

File: draft-ietf-quic-tls.md [hunk 2]
```diff
@@ -651,6 +651,17 @@ verification that the identity of the server is included in a certificate and
...
+  using keys with small public key representations, like ECDSA; or, by using
```
Comment (marten-seemann) [inline draft-ietf-quic-tls.md:662]: s/or/and/?

File: draft-ietf-quic-tls.md [hunk 3]
```diff
@@ -756,8 +767,8 @@ In TLS over TCP, the HelloRetryRequest feature (see Section 4.1.4 of
 well as for a stateless round-trip check. From the perspective of QUIC, this
 just looks like additional messages carried in Initial packets. Although it is
 in principle possible to use this feature for address verification in QUIC,
-QUIC implementations SHOULD instead use the Retry feature (see Section 8.1 of
-{{QUIC-TRANSPORT}}). HelloRetryRequest is still used to request key shares.
+QUIC implementations SHOULD instead use the Retry feature; see Section 8.1 of
```
Comment (marten-seemann) [inline draft-ietf-quic-tls.md:770]: This seems unrelated. Was this supposed to go in #3826?

File: draft-ietf-quic-tls.md (hunk 2 above)
Comment (martinthomson) [reply draft-ietf-quic-tls.md]: I think that a logical OR is sufficient here.  Though nothing stops someone from doing more than one, and maybe there are things that mean you need to (like being unwilling to reduce the number of cnames on certificates for $reasons.

File: draft-ietf-quic-tls.md [hunk 4]
```diff
@@ -651,6 +651,18 @@ verification that the identity of the server is included in a certificate and
...
 
+Note:
+
+: Where servers provide certificates for authentication, the size of
+  the certificate chain can consume a large number of bytes.  Controlling the
+  size of certificate chains is critical to performance in QUIC as servers are
+  limited to sending 3 bytes for every byte received prior to validating the
+  client address; see Section 8.1 of {{QUIC-TRANSPORT}}.  The size of a
+  certificate chain can be managed by limiting the number of names or
+  extensions; using keys with small public key representations, like ECDSA; or,
```
Comment (MikeBishop) [inline draft-ietf-quic-tls.md:662]:
```suggestion
  extensions; using keys with small public key representations, like ECDSA; or
```
